intelmq/bots/parsers/alienvault/parser_otx.py

```diff
@@ -33,50 +35,54 @@ def process(self):
+                            else 'http://' + indicator["indicator"]
+                        path = parse.urlparse(resource).path
+                        if len(path) > 0:
+                            event.add('source.url', resource)
+                        else:
+                            event.add('source.fqdn',
+                                      indicator["indicator"])
+                    # IP addresses
+                    elif indicator["type"] in ['IPv4', 'IPv6']:
+                        event.add('source.ip',
+                                  indicator["indicator"])
+                    # emails
+                    elif indicator["type"] == 'email':
+                        event.add('source.account',
+                                  indicator["indicator"])
+                    # URLs
+                    elif indicator["type"] in ['URL', 'URI']:
+                        resource = indicator["indicator"] \
+                            if '://' in indicator["indicator"] \
+                            else 'http://' + indicator["indicator"]
                         event.add('source.url', resource)
-                    else:
-                        event.add('source.fqdn',
+                    # CIDR
+                    elif indicator["type"] in ['CIDR']:
+                        event.add('source.network',
                                   indicator["indicator"])
-                # IP addresses
-                elif indicator["type"] in ['IPv4', 'IPv6']:
-                    event.add('source.ip',
-                              indicator["indicator"])
-                # emails
-                elif indicator["type"] == 'email':
-                    event.add('source.account',
-                              indicator["indicator"])
-                # URLs
-                elif indicator["type"] in ['URL', 'URI']:
-                    resource = indicator["indicator"] \
-                        if '://' in indicator["indicator"] \
-                        else 'http://' + indicator["indicator"]
-                    event.add('source.url', resource)
-                # CIDR
-                elif indicator["type"] in ['CIDR']:
-                    event.add('source.network',
-                              indicator["indicator"])
-
-                # CVE
-                elif indicator["type"] in ['CVE']:
-                    additional_indicator['CVE'] = indicator["indicator"]
-                    # TODO: Process these IoCs: FilePath, Mutex
-                else:
+
+                    # CVE
+                    elif indicator["type"] in ['CVE']:
+                        additional_indicator['CVE'] = indicator["indicator"]
+                        # TODO: Process these IoCs: FilePath, Mutex
+                    else:
+                        continue
+                except InvalidValue:
+                    self.logger.exception('Failed to parse line: {}\n{}'.format(indicator, format_exc()))
```

Comment:
Thanks, now I understand the dump functionality pretty better!

I extended the functionality of parse_line. In the docs, there is written `parse_line: Parses elements, returns an Event.` But till now, it handled list of Events only. Now, it can get list, the single event or None (no event in the row).
* Please comment if you find this extension useful!
* I've added methods for parsing JSON. What about to create JsonParserBot, CsvParserBot and CsvDictParserBot as children of ParserBot. Now, we have to redefine `parse` and `recover_line` if we want to use csv or csvDict. Maybe it would be easier if adopting CsvDictParserBot automatically used parse_csv_dict instead of parse_csv and recover_line_csv_dict instead of recover_line.